File: CoreLibrary/Package/GUI_Components/Controls/_Validation/InputtedValueValidation.ts
/* eslint-disable max-classes-per-file --
* This limitation is unsolicited for the namespaced classes, however, there is no ESLint option allowing this case. */
import { isBoolean, Logger } from "@yamato-daiwa/es-extensions";

import inputtedValueValidationLocalization__english from "./InputtedValueValidationLocaization.english";


abstract class InputtedValueValidation {

  public static localization: InputtedValueValidation.Localization = inputtedValueValidationLocalization__english;


  public readonly isInputRequired: InputtedValueValidation.InputRequirementChecker;
  public readonly hasValueBeenOmitted: InputtedValueValidation.OmittedValueChecker;

  protected readonly requiredInputIsMissingValidationErrorMessage: string;
  protected readonly staticRules: ReadonlyArray<InputtedValueValidation.Rule>;
  protected readonly contextDependentRules: ReadonlyArray<InputtedValueValidation.Rule>;
  protected readonly asynchronousRules: ReadonlyArray<InputtedValueValidation.AsynchronousRule>;


  protected constructor(compoundParameter: InputtedValueValidation.ConstructorCompoundParameter) {

    this.hasValueBeenOmitted = compoundParameter.omittedValueChecker;

    if (isBoolean(compoundParameter.isInputRequired)) {
      const isInputRequired: boolean = compoundParameter.isInputRequired;
      this.isInputRequired = (): boolean => isInputRequired;
    } else {
      this.isInputRequired = compoundParameter.isInputRequired;
    }

    this.requiredInputIsMissingValidationErrorMessage =
        compoundParameter.requiredInputIsMissingValidationErrorMessage ??
        InputtedValueValidation.localization.requiredInputIsMissingValidationErrorMessage;

    this.staticRules = compoundParameter.staticRules ?? [];
    this.contextDependentRules = compoundParameter.contextDependentRules ?? [];
    this.asynchronousRules = compoundParameter.asynchronousRules ?? [];

  }


  public validate(
    rawValue: unknown,
    {
      mustPostponeAsynchronousValidation = false,
      asynchronousChecksCallback,
      messagesOfExternallyDetectedValidationErrors = []
    }: Readonly<{
      mustPostponeAsynchronousValidation?: boolean;
      asynchronousChecksCallback?: InputtedValueValidation.AsynchronousChecks.Callback;
      messagesOfExternallyDetectedValidationErrors?: ReadonlyArray<string>;
    }> = {}
  ): InputtedValueValidation.Result {

    const isInputRequired: boolean = this.isInputRequired();

    if (this.hasValueBeenOmitted(rawValue)) {
      return isInputRequired ?
          {
            isValid: false,
            errorsMessages: [ this.requiredInputIsMissingValidationErrorMessage ]
          } :
          { isValid: true };
    }


    const validationErrorsMessages: Array<string> = [ ...messagesOfExternallyDetectedValidationErrors ];

    for (const staticValidationRule of this.staticRules) {

      const checkingResult: InputtedValueValidation.Rule.CheckingResult = staticValidationRule.check(rawValue);

      if (!checkingResult.isValid) {

        validationErrorsMessages.push(checkingResult.errorMessage);

        if (staticValidationRule.mustFinishValidationIfValueIsInvalid) {
          break;
        }

      }

    }

    if (validationErrorsMessages.length > 0) {
      return {
        isValid: false,
        errorsMessages: validationErrorsMessages
      };
    }


    for (const contextDependentValidationRule of this.contextDependentRules) {

      const checkingResult: InputtedValueValidation.Rule.CheckingResult = contextDependentValidationRule.check(rawValue);

      if (!checkingResult.isValid) {

        validationErrorsMessages.push(checkingResult.errorMessage);

        if (contextDependentValidationRule.mustFinishValidationIfValueIsInvalid) {
          break;
        }

      }

    }

    if (validationErrorsMessages.length > 0) {
      return {
        isValid: false,
        errorsMessages: validationErrorsMessages
      };
    }


    const validationResult: InputtedValueValidation.Result = { isValid: true };

    if (!mustPostponeAsynchronousValidation) {
      this.executeAsynchronousChecksIfAny(rawValue, validationResult, asynchronousChecksCallback);
    }


    return validationResult;

  }

  public executeAsynchronousChecksIfAny(
    rawValue: unknown,
    currentValidationResult: InputtedValueValidation.Result,
    asynchronousChecksCallback?: InputtedValueValidation.AsynchronousChecks.Callback
  ): void {

    if (this.asynchronousRules.length === 0) {
      return;
    }


    const asynchronousChecks: InputtedValueValidation.AsynchronousChecks = this.asynchronousRules.reduce(
      (
        accumulatingValue: InputtedValueValidation.AsynchronousChecks,
        asynchronousValidationRule: InputtedValueValidation.AsynchronousRule
      ): InputtedValueValidation.AsynchronousChecks => {

        accumulatingValue[asynchronousValidationRule.ID] = {
          message: asynchronousValidationRule.messages.checkingInProgress,
          isPending: true,
          hasInvalidValueBeenConfirmed: false,
          hasValidValueBeenConfirmed: false,
          hasErrorOccurred: false
        };

        return accumulatingValue;

      },
      {}
    );

    asynchronousChecksCallback?.(
      new InputtedValueValidation.AsynchronousChecks.Status(asynchronousChecks),
      currentValidationResult
    );

    for (const validationRule of this.asynchronousRules) {

      validationRule.

          check(rawValue).

          then((checkingResult: InputtedValueValidation.AsynchronousRule.CheckingResult): void => {

						asynchronousChecks[validationRule.ID] = {
							isPending: false,
							hasValidValueBeenConfirmed: checkingResult.isValid,
							hasInvalidValueBeenConfirmed: !checkingResult.isValid,
							hasErrorOccurred: false,
							message: checkingResult.isValid ?
									validationRule.messages.validValueHasBeenConfirmed :
									checkingResult.errorMessage ?? validationRule.messages.invalidValueHasBeenConfirmed
						};

            const asynchronousChecksStatus: InputtedValueValidation.AsynchronousChecks.Status =
                new InputtedValueValidation.AsynchronousChecks.Status(asynchronousChecks);

            const errorsMessages: ReadonlyArray<string> = [
              ...currentValidationResult.isValid ? [] : currentValidationResult.errorsMessages,
              ...asynchronousChecksStatus.errorsMessages
            ];

						asynchronousChecksCallback?.(
              asynchronousChecksStatus,
              {
                isValid: errorsMessages.length === 0,
                errorsMessages
              }
            );

					}).

          catch((error: unknown): void => {

            Logger.logError({
              errorType: "AsynchronousValidationFailure",
              title: "Asynchronous validation failed",
              description: `The asynchronous validation ${ validationRule.ID } has failed.`,
              occurrenceLocation: "inputtedValueValidation.executeAsynchronousChecksIfAny(rawValue)",
              caughtError: error
            });

            asynchronousChecks[validationRule.ID] = {
              isPending: false,
              hasValidValueBeenConfirmed: false,
              hasInvalidValueBeenConfirmed: false,
              hasErrorOccurred: true,
              message: validationRule.messages.errorHasOccurred
            };

            asynchronousChecksCallback?.(
              new InputtedValueValidation.AsynchronousChecks.Status(asynchronousChecks),
              currentValidationResult
            );

          });

    }

  }

}


namespace InputtedValueValidation {

  export type Localization = Readonly<{
    requiredInputIsMissingValidationErrorMessage: string;
  }>;


  export type ConstructorCompoundParameter = Readonly<{
    isInputRequired: boolean | InputRequirementChecker;
    omittedValueChecker: OmittedValueChecker;
    requiredInputIsMissingValidationErrorMessage?: string;
    staticRules?: ReadonlyArray<Rule>;
    contextDependentRules?: ReadonlyArray<Rule>;
    asynchronousRules?: ReadonlyArray<AsynchronousRule>;
    asynchronousValidationsCallback?: AsynchronousChecks.Callback;
    localization?: Localization;
  }>;


  export type OmittedValueChecker = (rawValue: unknown) => boolean;
  export type InputRequirementChecker = () => boolean;


  export interface Rule {

    readonly mustFinishValidationIfValueIsInvalid: boolean;

    readonly check: (rawValue: unknown) => Rule.CheckingResult;

  }

  export namespace Rule {

    export type CheckingResult = Readonly<
      {
        isValid: true;
      } | {
        isValid: false;
        errorMessage: string;
      }
    >;

    export type ConstructorParameter = Readonly<{
      mustFinishValidationIfValueIsInvalid?: boolean;
    }>;

  }


  export interface AsynchronousRule {
    readonly ID: string;
    readonly messages: AsynchronousRule.Messages;
    readonly check: (rawValue: unknown) => Promise<AsynchronousRule.CheckingResult>;
  }

  export namespace AsynchronousRule {

    export type Messages = Readonly<{
      checkingInProgress: string;
      validValueHasBeenConfirmed: string;
      invalidValueHasBeenConfirmed: string;
      errorHasOccurred: string;
    }>;

    export type CheckingResult = Readonly<{
      isValid: boolean;
      errorMessage?: string;
    }>;

  }


  export type Result = Readonly<
    {
      isValid: true;
    } |
    {
      isValid: false;
      errorsMessages: ReadonlyArray<string>;
    }
  >;


  export type AsynchronousChecks = { [validationRuleName: string]: AsynchronousCheck.Status; };

  export namespace AsynchronousChecks {

    export type Callback = (status: Status, currentValidationResult: Result) => void;

    export class Status {

      public readonly checks: AsynchronousChecks;
      public readonly hasAtLeastOneCheckNotFinishedYet: boolean = false;
      public readonly hasAllChecksFinishedWithAnyOutcome: boolean = true;
      public readonly hasAtLeastOneCheckErrorOccurred: boolean = false;
      public readonly hasNoInvalidValuesBeenConfirmed: boolean = true;
      public readonly hasAtLeastOneInvalidValueBeenConfirmed: boolean = false;
      public readonly errorsMessages: Array<string> = [];

      public constructor(checks: AsynchronousChecks) {

        this.checks = checks;

        for (const checking of Object.values(checks)) {

          if (checking.isPending) {
            this.hasAtLeastOneCheckNotFinishedYet = true;
            this.hasAllChecksFinishedWithAnyOutcome = false;
          }


          if (checking.hasErrorOccurred) {
            this.hasAtLeastOneCheckErrorOccurred = true;
          }

          if (checking.hasInvalidValueBeenConfirmed) {
            this.hasNoInvalidValuesBeenConfirmed = false;
            this.hasAtLeastOneInvalidValueBeenConfirmed = true;
            this.errorsMessages.push(checking.message);
          }

        }

      }

    }

  }

  export namespace AsynchronousCheck {

    export type Status = {
      message: string;
      isPending: boolean;
      hasValidValueBeenConfirmed: boolean;
      hasInvalidValueBeenConfirmed: boolean;
      hasErrorOccurred: boolean;
    };

  }

}


export default InputtedValueValidation;

File: AdaptationsToFrameworks/Blazor/Package/GUI_Components/Controls/Validation/InputtedValueValidation.cs
using System.Diagnostics;


namespace YamatoDaiwa.Frontend.GUI_Components.Controls.Validation;


public abstract class InputtedValueValidation
{

  public interface ILocalization
  {
    public string RequiredInputIsMissingValidationErrorMessage { get; }
  }

  public static ILocalization Localization = new InputtedValueValidationEnglishLocalization();

  
  public readonly Func<bool> IsInputtingRequirementChecker;
  public readonly Func<object?, bool> HasValueBeenOmitted;
  
  protected readonly string requiredInputIsMissingValidationErrorMessage;

  
  public interface IRule
  {
  
    public bool MustFinishValidationIfValueIsInvalid { get; init; }
  
    public IRule.CheckingResult Check(object rawValue);

    public struct CheckingResult
    {
      public string? ErrorMessage { get; init; }
      public bool IsValid => this.ErrorMessage is not null;
    }
  
  }
  
  public interface IAsynchronousRule
  {
  
    public string ID { get; init; }
    public IAsynchronousRule.Messages messages { get; init; }
    
    public Task<CheckingResult> Check(object rawValue);
    
    public struct Messages
    {
      public string CheckingInProgress { get; internal set; }
      public string ValidValueHasBeenConfirmed { get; internal set; }
      public string InvalidValueHasBeenConfirmed { get; internal set; }
      public string ErrorHasOccurred { get; internal set; }
    }
    
    public struct CheckingResult
    {
      public string? ErrorMessage { get; init; }
      public bool IsValid => this.ErrorMessage is not null;
    }
  
  }
  
  protected IRule[] StaticRules;
  protected IRule[] ContextDependentRules;
  protected IAsynchronousRule[] AsynchronousRules;
  
  
  /* ━━━ Constructor ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  protected InputtedValueValidation(
    Func<object?, bool> omittedValueChecker,
    bool? isInputRequired = null,
    Func<bool>? inputtingRequirementChecker = null,
    string? requiredInputIsMissingValidationErrorMessage = null,
    IRule[]? staticRules = null,
    IRule[]? contextDependentRules = null,
    IAsynchronousRule[]? asynchronousRules = null 
  )
  {
    
    this.HasValueBeenOmitted = omittedValueChecker;

    if (isInputRequired is not null)
    {
      
      if (inputtingRequirementChecker is not null)
      {
        throw new ArgumentException(
          "The \"isInputRequired\" and \"inputtingRequirementChecker\" parameters are incompatible. " +
          "Please specify one of them."
        );
      }
      
      
      this.IsInputtingRequirementChecker = () => (bool)isInputRequired;
      
    } else if (inputtingRequirementChecker is not null)
    {
      this.IsInputtingRequirementChecker = inputtingRequirementChecker;
    } else
    {
      throw new ArgumentException(
      "Either \"isInputRequired\" or \"inputtingRequirementChecker\" must be specified but not both."
      );
    }
    
    this.requiredInputIsMissingValidationErrorMessage =
        requiredInputIsMissingValidationErrorMessage ??
        InputtedValueValidation.Localization.RequiredInputIsMissingValidationErrorMessage;

    
    this.StaticRules = staticRules ?? Array.Empty<IRule>();
    this.ContextDependentRules = contextDependentRules ?? Array.Empty<IRule>();
    this.AsynchronousRules = asynchronousRules ?? Array.Empty<IAsynchronousRule>();

  }
  
  
  public struct Result
  {
    public string[] ErrorsMessages { get; init; }
    public bool IsValid => this.ErrorsMessages.Length > 0;
  }
  
  public Result Validate(
    object rawValue,
    bool mustPostponeAsynchronousValidation = false,
    Action<AsynchronousChecks.Status, Result>? asynchronousChecksCallback = null,
    string[]? messagesOfExternallyDetectedValidationErrors = null
  )
  {

    bool isInputRequired = this.IsInputtingRequirementChecker();
    
    if (this.HasValueBeenOmitted(rawValue))
    {
      return new Result
      {
        ErrorsMessages = isInputRequired ? [ this.requiredInputIsMissingValidationErrorMessage ] : [] 
      };
    }
    
    
    List<string> validationErrorsMessages = messagesOfExternallyDetectedValidationErrors?.ToList() ?? [];

    foreach (IRule staticValidationRule in this.StaticRules)
    {

      IRule.CheckingResult checkingCheckingResult = staticValidationRule.Check(rawValue);

      if (checkingCheckingResult.ErrorMessage is not null)
      {

        validationErrorsMessages.Add(checkingCheckingResult.ErrorMessage);

        if (staticValidationRule.MustFinishValidationIfValueIsInvalid)
        {
          break;
        }

      }
      
    }

    if (validationErrorsMessages.Count > 0)
    {
      return new Result
      {
        ErrorsMessages = validationErrorsMessages.ToArray()
      };
    }
    
    
    foreach (IRule contextDependentValidationRule in this.ContextDependentRules)
    {

      IRule.CheckingResult checkingCheckingResult = contextDependentValidationRule.Check(rawValue);

      if (checkingCheckingResult.ErrorMessage is not null)
      {

        validationErrorsMessages.Add(checkingCheckingResult.ErrorMessage);

        if (contextDependentValidationRule.MustFinishValidationIfValueIsInvalid)
        {
          break;
        }

      }
      
    }
    
    if (validationErrorsMessages.Count > 0)
    {
      return new Result
      {
        ErrorsMessages = validationErrorsMessages.ToArray()
      };
    }


    Result validationResult = new Result { ErrorsMessages = Array.Empty<string>() };
 
    
    if (!mustPostponeAsynchronousValidation)
    {
      this.executeAsynchronousChecksIfAny(rawValue, validationResult, asynchronousChecksCallback); 
    }
    
    return validationResult;

  }


  /* ━━━ Asynchronous validation ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  public abstract class AsynchronousChecks
  {

    public readonly struct Status
    {

      public readonly Dictionary<string, AsynchronousCheck.Status> Checks;
      public readonly bool HasAtLeastOneCheckNotFinishedYet = false;
      public readonly bool HasAllChecksFinishedWithAnyOutcome = true;
      public readonly bool HasAtLeastOneCheckErrorOccurred = false;
      public readonly bool HasNoInvalidValuesBeenConfirmed = true;
      public readonly bool HasAtLeastOneInvalidValueBeenConfirmed = false;
      public readonly string[] ErrorsMessages;

      public Status(Dictionary<string, AsynchronousCheck.Status> checks)
      {
        
        this.Checks = checks;

        List<string> errorsMessages = new();

        foreach ((string _, AsynchronousCheck.Status checking) in checks)
        {

          if (checking.IsPending)
          {
            this.HasAtLeastOneCheckNotFinishedYet = true;
            this.HasAllChecksFinishedWithAnyOutcome = false;
          }


          if (checking.HasErrorOccurred)
          {
            this.HasAtLeastOneCheckErrorOccurred = true;
          }


          if (checking.HasInvalidValueBeenConfirmed)
          {
            this.HasNoInvalidValuesBeenConfirmed = false;
            this.HasAtLeastOneInvalidValueBeenConfirmed = true;
            errorsMessages.Add(checking.Message);
          }

        }
        
        this.ErrorsMessages = errorsMessages.ToArray();
        
      }
    }
    
  }
  
  
  public abstract class AsynchronousCheck
  {

    public struct Status
    {
      public required string Message { get; set; }
      public required bool IsPending { get; set; }
      public required bool HasValidValueBeenConfirmed { get; set; } 
      public required bool HasInvalidValueBeenConfirmed { get; set; }
      public required bool HasErrorOccurred { get; set; }
    }
    
  }
  
  
  public void executeAsynchronousChecksIfAny(
    object rawValue, 
    Result currentValidationResult, 
    Action<AsynchronousChecks.Status, Result>? asynchronousChecksCallback
  )
  {

    if (this.AsynchronousRules.Length == 0)
    {
      return;
    }


    Dictionary<string, AsynchronousCheck.Status> asynchronousChecks = this.AsynchronousRules.
        ToDictionary(
          asynchronousRule => asynchronousRule.ID, 
          asynchronousRule => new AsynchronousCheck.Status
          {
            Message = asynchronousRule.messages.CheckingInProgress,
            IsPending = true,
            HasInvalidValueBeenConfirmed = false,
            HasValidValueBeenConfirmed = false,
            HasErrorOccurred = false
          }
        );

    asynchronousChecksCallback?.Invoke(new AsynchronousChecks.Status(asynchronousChecks), currentValidationResult);

    IEnumerable<Task> validationTasks = this.AsynchronousRules.Select(
      async validationRule =>
      {

        IAsynchronousRule.CheckingResult checkingResult;
        
        try
        {

          checkingResult = await validationRule.Check(rawValue);

        }
        catch (Exception exception)
        {

          Debug.WriteLine(
            "AsynchronousValidationFailedException: Asynchronous validation failed\n" +
            $"The asynchronous validation { validationRule.ID } has failed.\n" +
            new System.Diagnostics.StackTrace()
          );
          
          asynchronousChecks[validationRule.ID] = new AsynchronousCheck.Status
          {
            IsPending = false,
            HasValidValueBeenConfirmed = false,
            HasInvalidValueBeenConfirmed = false,
            HasErrorOccurred = true,
            Message = validationRule.messages.ErrorHasOccurred
          };
          
          asynchronousChecksCallback?.Invoke(new AsynchronousChecks.Status(asynchronousChecks), currentValidationResult);
          
          return;
          
        }
        
        
        asynchronousChecks[validationRule.ID] = new AsynchronousCheck.Status
        {
          IsPending = false,
          HasValidValueBeenConfirmed = checkingResult.IsValid,
          HasInvalidValueBeenConfirmed = !checkingResult.IsValid,
          HasErrorOccurred = false,
          Message = checkingResult.IsValid ? 
              validationRule.messages.ValidValueHasBeenConfirmed :
              checkingResult.ErrorMessage ?? validationRule.messages.InvalidValueHasBeenConfirmed 
        };
        
        AsynchronousChecks.Status asynchronousChecksStatus = new(asynchronousChecks);

        string[] errorsMessages = currentValidationResult.ErrorsMessages.
            Concat(asynchronousChecksStatus.ErrorsMessages).
            ToArray();

        asynchronousChecksCallback?.Invoke(asynchronousChecksStatus, new Result { ErrorsMessages = errorsMessages });

      }
    );

    Task.WhenAll(validationTasks);

  }
  
}